### PR TITLE
fix: improve autocompletion hints discovery

### LINKED ISSUE #17684: [Feature]: Deprioritise ranking of JS constructors on autocomplete results
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

function autocomplete

Currently

1. The best match result autocompletes the Function() constructor.
2. The second result autocompletes the function keyword
3. The last result autocompletes the function name(params){    } statement

Expected order
1. The best match result autocompletes the function name(params){    } statement.
2. The second result autocompletes the function keyword
3. The last result autocompletes the Function() constructor (as this is the least frequently used)

### Steps To Reproduce

1. Go to any code editor field
2. type `{{fun}}`
3. See Autocompletion order

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #27870: Sort autocomplete results based on recency of usage
Results should be sorted based on recency and context (for instance, a query or local variable that I just created should show up at the top)

### PR DESCRIPTION
**Pull request by ohansFavour**
## Description

This PR improves autocompletion hints discovery by

- Taking entities' recency of usage into consideration when sorting hints
- Showing entity names at the top, before supported functions and properties
- Deprioritizing the Function constructor and the MainContainer entity

#### PR fixes following issue(s)
Fixes #27870 
Fixes #17684 
Fixes https://github.com/appsmithorg/appsmith/issues/24975

#### Type of change
- New feature (non-breaking change which adds functionality)
 
- [ ] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Test Plan
- [x] check for autocomplete results when multiple widgets of the same type are used
- [x] check for entity renaming behavior
- [x] verify function and mainContainer have been removed
- [x] verify that recent entity changes affects autocomplete ranking
>
>
#### Issues raised during DP testing
https://github.com/appsmithorg/appsmith/pull/28222#issuecomment-1777487126
>
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [x] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [x] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [x] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

**Comment by ramsaptami** [comment]
lgtm
cc: @ohansFavour

File: app/client/src/utils/autocomplete/CodemirrorTernService.ts [hunk 1]
```diff
@@ -129,6 +131,28 @@ export function typeToIcon(type: string, isKeyword: boolean) {
   return cls + "completion " + cls + "completion-" + suffix;
 }
 
+function getRecencyWeight(
+  completion:
+    | string
```
Comment (arunvjn) [inline app/client/src/utils/autocomplete/CodemirrorTernService.ts:136]: Can this not be a part of the Sort Rule Engine ? Are there any challenges passing recentEntities to the sorter ?

File: app/client/src/utils/autocomplete/AutocompleteSortRules.ts [hunk 2]
```diff
@@ -249,16 +289,19 @@ class PriorityMatchRule implements AutocompleteRule {
...
  * Min score - 0
  */
 class ScopeMatchRule implements AutocompleteRule {
   static threshold = 1 << RuleWeight.ScopeMatch;
   computeScore(completion: Completion<TernCompletionResult>): number {
     let score = 0;
-    if (completion.origin === "[doc]" || completion.origin === "customDataTree")
-      score += PriorityMatchRule.threshold;
+    if (
+      completion.origin?.startsWith("[doc") ||
```
Comment (arunvjn) [inline app/client/src/utils/autocomplete/AutocompleteSortRules.ts:301]: Can you help me understand the pattern of `completion.origin` we are getting here ?

Comment (ohansFavour) [reply app/client/src/utils/autocomplete/AutocompleteSortRules.ts:301]: From [here](https://github.com/appsmithorg/appsmith/blob/d80fa0aeee660af3434db43bb486c9f4ade0cc57/app/client/src/utils/autocomplete/CodemirrorTernService.ts#L479), there are cases where the origin is "[doc1]" etc.

**Review by arunvjn — APPROVED** [review]
Looks good!

File: app/client/src/utils/autocomplete/CodemirrorTernService.ts (hunk 1 above)
Comment (ohansFavour) [reply app/client/src/utils/autocomplete/CodemirrorTernService.ts:136]: This deviates from the current pattern today. All of the metadata required to sort completions are currently embedded within the completions. 

Also, passing recentEntities to the sorter will require that it is passed to all of the rule sorting classes.

Comment (arunvjn) [reply app/client/src/utils/autocomplete/CodemirrorTernService.ts:136]: Yes makes sense!

**Pull request merged**